rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -14,391 +14,529 @@
 
 #include "rmw_context_impl_s.hpp"
 
+#include <chrono>
+#include <cstddef>
+#include <cstdint>
+#include <memory>
 #include <mutex>
 #include <optional>
+#include <stdexcept>
 #include <string>
 #include <thread>
+#include <unordered_map>
 #include <utility>
 
+#include "graph_cache.hpp"
 #include "guard_condition.hpp"
 #include "identifier.hpp"
-#include "liveliness_utils.hpp"
 #include "logging_macros.hpp"
+#include "rmw_node_data.hpp"
 #include "zenoh_config.hpp"
 #include "zenoh_router_check.hpp"
 
 #include "rcpputils/scope_exit.hpp"
 #include "rmw/error_handling.h"
-#include "rmw/impl/cpp/macros.hpp"
 
 // Megabytes of SHM to reserve.
 // TODO(clalancette): Make this configurable, or get it from the configuration
 #define SHM_BUFFER_SIZE_MB 10
 
-///=============================================================================
-void rmw_context_impl_s::graph_sub_data_handler(const z_sample_t * sample, void * data)
+// This global mapping of raw Data pointers to Data shared pointers allows graph_sub_data_handler()
+// to lookup the pointer, and gain a reference to a shared_ptr if it exists.
+// This guarantees that the Data object will not be destroyed while we are using it.
+static std::mutex data_to_data_shared_ptr_map_mutex;
+static std::unordered_map<Data *, std::shared_ptr<Data>> data_to_data_shared_ptr_map;
+
+static void graph_sub_data_handler(const z_sample_t * sample, void * data);
```

Comment:
This has to be here, as far as I understand.  The problem is that `class Data` refers to `graph_sub_data_handler`, and `graph_sub_data_handler` refers to `class Data`.  The forward declaration from the header file isn't sufficient, since that is only a partial type, and `graph_sub_data_handler` needs to know the full type to proceed.  So we break the circular dependency by forward-declaring it here.

Reply:
Thanks for clarifying!